### PR TITLE
Alpha TSG client

### PR DESCRIPTION
**Pull request by cheapRoc**
At the moment this is required to demo and work on authentication within TSG. I don't believe this is something we want to merge straight into master but there are a fix or two that I can peel off for logic already in master.

The goal was to present something that would hit TSG's templates endpoint and pass along a signed HTTP request we can re-purpose for authenticating a user within TSG on behalf of CloudAPI.

WIP ~ DO NOT MERGE

### CODE REVIEW

**Comment by cheapRoc** [comment]
I've opened #130 from off thread discussion around TSG endpoint configuration.

File: services/templates.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+package services
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+
```
Comment (sean-) [inline services/templates.go]: struct field AccountId should be AccountID

File: services/templates.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	"github.com/joyent/triton-go/client"
+	pkgerrors "github.com/pkg/errors"
```
Comment (sean-) [inline services/templates.go]: struct field ImageId should be ImageID

File: services/templates.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	fullPath := path.Join("/v1/tsg/templates", input.Name)
+	reqInputs := client.RequestInput{
+		Method: http.MethodGet,
+		Path:   fullPath,
+	}
+	respReader, err := c.client.ExecuteRequestTSG(ctx, reqInputs)
+	if respReader != nil {
+		defer respReader.Close()
```
Comment (sean-) [inline services/templates.go]: struct field AccountId should be AccountID

File: services/templates.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	}
+	if err != nil {
```
Comment (sean-) [inline services/templates.go]: struct field ImageId should be ImageID

**Comment by cheapRoc** [comment]
For this to be useful it currently needs all open PRs under joyent/triton-service-groups as well as joyent/triton-service-groups#7 and joyent/triton-service-groups#8.

File: services/templates.go (hunk 1 above)
Comment (cheapRoc) [reply services/templates.go]: Yeah I know, we need to change `AccountId` within [joyent/triton-service-groups](http://github.com/joyent/triton-service-groups) first.

Comment (cheapRoc) [reply services/templates.go]: Actually, with joyent/triton-service-groups#13 merged I can just use struct tags for the JSON parsing and rename to `AccountID` just on the `triton-go` side.

@stack72 opened joyent/triton-service-groups#14 for the renaming.

File: services/templates.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	"net/http"
+	"path"
+
+	"github.com/joyent/triton-go/client"
+	pkgerrors "github.com/pkg/errors"
+)
+
+const templatesPath = "/v1/tsg/templates"
+
+type TemplatesClient struct {
```
Comment (sean-) [inline services/templates.go]: struct field AccountId should be AccountID

File: services/templates.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+	client *client.Client
+}
```
Comment (sean-) [inline services/templates.go]: struct field ImageId should be ImageID

**Comment by cheapRoc** [comment]
Ok, we're all good on renames and JSON tags.

**Comment by cheapRoc** [comment]
I've rebased and removed the static services URL. There is now a manual override for local development (`TRITON_TSG_URL=http://localhost:3000/` env var). There's also URL generation based on the currently selected DC (`TRITON_URL`) for when the API goes live. This transforms `https?://(.*).api.joyent.com` into `https://tsg.%s.svc.joyent.zone`. This all means you can configure a `services` client by providing the same client configuration/authentication as you would today with `triton-go` (even if TSG doesn't include auth yet).

Examples have been updated under `examples/services/` and pair with latest TSG master ([3d7c120](https://github.com/joyent/triton-service-groups/commit/3d7c120eec951e9b11c44fc36a73821f68f95801)).

**Comment by cheapRoc** [comment]
Going to close and re-open with the branch I'm actually using for TF integration.

**Pull request closed without merging**